### PR TITLE
Improvements to overloading resolution code.

### PR DESCRIPTION
**Pull request by paulp**
Review by @retronym.

### CODE REVIEW

**Comment by paulp** [comment]
I forgot to mention: this increases the uniformity of followApply, but not all the way.

**Comment by paulp** [comment]
I borked a test with a last minute change; don't let that lone failure distract you, it's all quite sound.

**Comment by retronym** [comment]
Looks like a big improvement, give me a day or two to review it in detail.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 1]
```diff
@@ -49,6 +49,24 @@ trait Infer extends Checkable {
...
+    def improves(sym1: Symbol, sym2: Symbol) = (
+         sym2 == NoSymbol
+      || sym2.isError
+      || sym2.hasAnnotation(BridgeClass)
+      || isBetter(sym1, sym2)
+    )
+
+    alternatives sortWith improves match {
+      case best :: rest if rest.nonEmpty => best :: rest.filterNot(alt => improves(best, alt))
+      case bests                         => bests
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:66]: Wouldn't best :: Nil be more readable?

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 2]
```diff
@@ -672,6 +679,35 @@ trait Infer extends Checkable {
...
+        val simpleMatch   = paramsCount == argsCount
+        def varargsTarget = isVarArgsList(tpe.params)
+        def varargsMatch  = varargsTarget && (paramsCount - 1) <= argsCount
+        def tuplingMatch  = tuplingAllowed && (argsCount != 1) && (paramsCount == 1 || paramsCount == 2 && varargsTarget)
+
+        // A varargs star call, e.g. (x, y:_*) can only match a varargs method
+        // with the same number of parameters.  See SI-5859 for an example of what
+        // would fail were this not enforced before we arrived at isApplicable.
+        if (varargsStar)
+          varargsTarget && (paramsCount == argsCount)
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: paramsCount == argsCount ==> simpleMatch

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala [hunk 3]
```diff
@@ -1489,52 +1539,29 @@ trait Infer extends Checkable {
...
-          tree.setSymbol(best).setType(pre.memberType(best))
+          (    tp2 == ErrorType
+            || (!isWeaklyCompatible(tp2, pt) && isWeaklyCompatible(tp1, pt))
+            || isStrictlyMoreSpecific(tp1, tp2, sym1, sym2)
+          )
+        }
+        bests match {
+          case best :: Nil                              => tree setSymbol best setType (pre memberType best)
+          case best :: competing :: _ if alts0.nonEmpty => if (!pt.isErroneous) AmbiguousExprAlternativeError(tree, pre, best, competing, pt, isSecondTry)
+          case _                                        => if (bests.isEmpty || alts0.isEmpty) NoBestExprAlternativeError(tree, pt, isSecondTry) // todo: missing test case
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: the //todo is actually a bit misleading. we are missing test only for bests.isEmpty.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 1 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:66]: I'm not sure which case you mean, but neither of those cases simplifies to best :: Nil.  If you mean the second case, it also includes Nil (which does come through in practice.)

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 2 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: Good point!

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 3 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala]: My fault; I knew that was a lossy comment move when I did it.

File: src/compiler/scala/tools/nsc/typechecker/Infer.scala (hunk 1 above)
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Infer.scala:66]: True, got confused by the rest.nonEmpty. Looks correct.

**Comment by hubertp** [comment]
I wasn't assigned as a reviewer of this one but I spent my share of time in those methods and this is definitely an improvement. so lgtm.

**Comment by paulp** [comment]
Thanks.  Clearly I should be shopping at your review store more often.

**Comment by paulp** [comment]
@retronym, I am still keen for any input you would like to offer, but I am even more keen not to let my pull requests stack up when I have this lgtm and a pair of jenkins successes burning a hole in my pocket.  In other words, I'm merging.

**Pull request merged**